### PR TITLE
modules: hal_silabs: Add Silabs Gecko HAL for EFR32BG22 SoCs

### PR DESCRIPTION
**Pull request by sateeshkotapati**
Origin: Silicon Labs Gecko SDK
URL:
https://www.silabs.com/products/development-tools/software/simplicity-studio
Version: 3.1.2
Purpose: Add support for Silicon Labs EFR32BG22 SoCs
License: Zlib
Maintained-by: External

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

### CODE REVIEW

**Comment by sateeshkotapati** [comment]
Support added for Silabs' EFR32BG22 SoCs, which I would like to contribute to zephyr

**Comment by galak** [comment]
Please submit an associated PR on the Zephyr repo that adds SoC and board support that utilizes this.

**Review by chrta — CHANGES_REQUESTED** [review]
Thank you for your contribution, i just mark this as "Request changes" to block merging. I will do a proper review once the complete code is available, that adds support for this soc to zephyr.

It seems that the board initialization is now used from the service folder. This should be done in a consistent manner for all Silabs EXX32 SoCs (imho).
The README file https://github.com/zephyrproject-rtos/hal_silabs/blob/master/gecko/README should also be adapted to reflect what folders/files need to be copied.

**Comment by sateeshkotapati** [comment]
I missed gecko/README file in my first commit in this PR. That is updated now.
Board initialization is kept exclusive for EFR32BG22 SoCs and is not extended to all EXX32 devices because validation effort is huge for all the devices.

**Review by chrta — CHANGES_REQUESTED** [review]
Thank you for the updates. I am ok with keeping the currently supported EXX32 boards as they are now, i agree that it would be a lot of effort. My main points in the comments:

- It seems that my sdk version 3.1.2 i downloaded differs slightly from the code in this pr. This should be fixed/documented.
- Changes applied to the sdk should be done in a separate commit outlining the reasoning for this, e.g. adding a file sl_board_control_config.h. There should be a documentation why this is necessary and how the file was created etc.
- The newly added files in the service folder should be compiled if the SOC selects them and should not depend the SOC define itself. For this new CONFIG_-defines should be added.

File: gecko/CMakeLists.txt [hunk 1]
```diff
@@ -25,6 +39,24 @@ zephyr_compile_definitions(
...
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   service/device_init/src/sl_device_init_lfxo_s2.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   service/device_init/src/sl_device_init_clocks.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   service/device_init/src/sl_device_init_nvic.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   board/src/sl_board_init.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   service/power_manager/src/sl_power_manager.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   service/power_manager/src/sl_power_manager_hal_s0_s1.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   service/power_manager/src/sl_power_manager_hal_s2.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   service/hfxo_manager/src/sl_hfxo_manager.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   service/hfxo_manager/src/sl_hfxo_manager_hal_s2.c)
+zephyr_sources_ifdef(CONFIG_SOC_SERIES_EFR32BG22   common/src/sl_slist.c)
```
[the remark this answers is not in the document]
Comment (sateeshkotapati) [reply gecko/CMakeLists.txt]: Thank you for the review and detailed comments. I have incorporated all the review comments.

File: gecko/CMakeLists.txt [hunk 2]
```diff
@@ -37,14 +69,17 @@ zephyr_sources_ifdef(CONFIG_SOC_GECKO_PRS          emlib/src/em_prs.c)
 zephyr_sources_ifdef(CONFIG_SOC_GECKO_RMU          emlib/src/em_rmu.c)
 zephyr_sources_ifdef(CONFIG_SOC_GECKO_RTC          emlib/src/em_rtc.c)
 zephyr_sources_ifdef(CONFIG_SOC_GECKO_RTCC         emlib/src/em_rtcc.c)
+zephyr_sources_ifdef(CONFIG_SOC_GECKO_RTCC         service/sleeptimer/src/sl_sleeptimer_hal_rtcc.c)
+zephyr_sources_ifdef(CONFIG_SOC_GECKO_RTCC         service/sleeptimer/src/sl_sleeptimer.c)
```
[the remark this answers is not in the document]
Comment (sateeshkotapati) [reply gecko/CMakeLists.txt]: I have incorporated this review comment.

File: gecko/service/sleeptimer/test/main.c [hunk 3]
```diff
@@ -0,0 +1,2441 @@
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (sateeshkotapati) [reply gecko/service/sleeptimer/test/main.c]: All test directories which are not used are removed.

File: gecko/board/inc/sl_board_control_config.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
+#ifndef SL_BOARD_CONTROL_CONFIG_H
```
[the remark this answers is not in the document]
Comment (sateeshkotapati) [reply gecko/board/inc/sl_board_control_config.h:1]: This file is copied from hardware\board\config directory. This is required as this header file is included in sl_board_init.c file.

File: gecko/board/src/sl_board_init.c [hunk 5]
```diff
@@ -0,0 +1,98 @@
+/***************************************************************************//**
```
[the remark this answers is not in the document]
Comment (sateeshkotapati) [reply gecko/board/src/sl_board_init.c:1]: Updated.

File: gecko/common/inc/sl_status.h [hunk 6]
```diff
@@ -0,0 +1,425 @@
+/*******************************************************************************
```
[the remark this answers is not in the document]
Comment (sateeshkotapati) [reply gecko/common/inc/sl_status.h:1]: These are necessary as these are used in power_manager files. Updated the files.

File: gecko/Device/SiliconLabs/EFR32BG22/Include/efr32bg22c112f352gm40.h [hunk 7]
```diff
@@ -0,0 +1,1252 @@
+/**************************************************************************//**
+ * @file
+ * @brief CMSIS Cortex-M Peripheral Access Layer Header File
+ *        for EFR32BG22C112F352GM40
+ ******************************************************************************
+ * # License
+ * <b>Copyright 2020 Silicon Laboratories, Inc. www.silabs.com</b>
```
[the remark this answers is not in the document]
Comment (sateeshkotapati) [reply gecko/Device/SiliconLabs/EFR32BG22/Include/efr32bg22c112f352gm40.h]: updated.

File: gecko/board/inc/sl_board_control_config.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrta) [reply gecko/board/inc/sl_board_control_config.h:1]: 1. Can you document this in the README file -> what to do if a new board is added
2. How does this work if a second board should be supported in a similar way that needs a different config?

**Review by chrta — CHANGES_REQUESTED** [review]
Thank you for the changes. My open issues:
1. Please document in the README file how the file sl_board_control_config.h is generated, if a new board should be added.
2. How does it work if a second board is added that needs a different sl_board_control_config.h
3. Please squash the commits. I would only expect 2 commits (or maybe only one). One adding the device etc to this repository and a second commit that contains the necessary manual changes/additions.

**Review by chrta — CHANGES_REQUESTED** [review]
Thank you for the changes, especially the addition to the README file. The only open issue/question i have is:

How does it work if a second board is added that needs a different configuration, i.e. sl_board_control_config.h. I think currently it does not work, but imho it should work.

I would prefer if the generated files would be in different subfolders containing the board name in the path and the build system takes care that the correct files are compiled and inclusion paths are set correctly.

For this one board the current approach might work. But once a second board is added it does not work without manually modifying the generated files in this repository. Imho this is bad.

**Comment by sateeshkotapati** [comment]
> Thank you for the changes, especially the addition to the README file. The only open issue/question i have is:
> 
> How does it work if a second board is added that needs a different configuration, i.e. sl_board_control_config.h. I think currently it does not work, but imho it should work.
> 
> I would prefer if the generated files would be in different subfolders containing the board name in the path and the build system takes care that the correct files are compiled and inclusion paths are set correctly.
> 
> For this one board the current approach might work. But once a second board is added it does not work without manually modifying the generated files in this repository. Imho this is bad.

I have added the configuration files for all the silabs devices under gecko/board/config. sl_board_control_config.h will be chosen based on the board name that is given in zephyr\boards\arm\efr32_radio\Kconfig.defconfig

**Comment by chrta** [comment]
> I have added the configuration files for all the silabs devices under gecko/board/config. sl_board_control_config.h will be chosen based on the board name that is given in zephyr\boards\arm\efr32_radio\Kconfig.defconfig

Oh yes, sorry i missed that. I noticed that you modified the board name in https://github.com/zephyrproject-rtos/zephyr/pull/35361/files#diff-e63781934d96e8af747cf94696a9ba7d1f946e4c0113aed1969212d8c4075e35R11 so it does not include the prefix `efr32_radio_` like all the other boards do. I would prefer if It would have the prefix and stripping it is handled in gecko/CMakeLists.txt in a similar way as the cmake variable SILABS_GECKO_DEVICE is created. E.g. including this:

```diff
diff --git a/gecko/CMakeLists.txt b/gecko/CMakeLists.txt
index 731c93c..66f99a3 100644
--- a/gecko/CMakeLists.txt
+++ b/gecko/CMakeLists.txt
@@ -9,6 +9,15 @@
 # respectively.
 string(TOUPPER ${CONFIG_SOC_SERIES} SILABS_GECKO_DEVICE)
 
+# Remove any prefix from the board, e.g. efr32_radio_
+string(FIND ${CONFIG_BOARD} "brd" SILABS_GECKO_BRD_INDEX)
+
+if (${SILABS_GECKO_BRD_INDEX} LESS 0)
+  set(SILABS_GECKO_BOARD ${CONFIG_BOARD})
+else()
+  string(SUBSTRING ${CONFIG_BOARD} ${SILABS_GECKO_BRD_INDEX} -1 SILABS_GECKO_BOARD)
+endif()
+
 set(SILABS_GECKO_PART_NUMBER ${CONFIG_SOC_PART_NUMBER})
 
 zephyr_include_directories(
@@ -28,7 +37,7 @@ zephyr_include_directories(
   service/device_init/config/s2
   service/device_init/config/s2/sdid205
   board/inc
-  board/config/${CONFIG_BOARD}
+  board/config/${SILABS_GECKO_BOARD}
   )
 
 # The gecko SDK uses the cpu name to include the matching device header.
```

This way CONFIG_BOARD for the new boards can include prefixes.
I think for other boards where the config folder name is e.g. gecko/board/config/brd4100a_brd4001a this should be handled in this repository (hal_silabs) by mofifying gecko/CMakeLists.txt and not by modifying CONFIG_BOARD in the main zephyr repository.

**Review by chrta — APPROVED** [review]
Thank you, lgtm

File: gecko/CMakeLists.txt [hunk 8]
```diff
@@ -20,6 +20,10 @@ endif()
 
 set(SILABS_GECKO_PART_NUMBER ${CONFIG_SOC_PART_NUMBER})
 
+if (${CONFIG_BOARD} STREQUAL "efr32bg_sltb010a")
+  set(SILABS_GECKO_BOARD "brd4184b")
```
Comment (galak) [inline gecko/CMakeLists.txt:24]: What is SILABS_GECKO_BOARD used for?

Comment (sateeshkotapati) [reply gecko/CMakeLists.txt:24]: Sorry for updating late. I missed the notification.
It is used to get board specific configuration from directory "zephyrproject\modules\hal\silabs\gecko\board\config"

Comment (galak) [reply gecko/CMakeLists.txt:24]: What are you using from board/config/*?

**Comment by galak** [comment]
Can you split out the code import from any changes/additions as different commits.  So the CMakeLists.txt would be a separate commit for example.

**Review by galak — CHANGES_REQUESTED** [review]
Would prefer not to import the boards/configs.  I'm also unsure about the services/.  Can we limit the initial import / support to  Device/SiliconLabs/EFR32BG22?

**Review by chrta — COMMENTED** [review]
I remove my approval until discussion is resolved. Also a lot changed since my last review.

**Comment by galak** [comment]
Noticed another issue with some of the board/ files is they are not licensed under a license that would be acceptable to the Zephyr project.

https://docs.zephyrproject.org/latest/contribute/external.html#software-license

**Comment by sateeshkotapati** [comment]
Please review this PR, Thanks

**Comment by sateeshkotapati** [comment]
Hi Zephyr community, 
Please re-review this PR. I have incorporated all the previous review comments.
Our goal to add more Silabs devices is blocked on this PR. 
Thank you and appreciate your time. 

Regards,
Sateesh

**Comment by kgugala** [comment]
this PR is deprecated. The changes have been merged in the followup PRs. Closing this one

**Pull request closed without merging**